step/step.go

```diff
@@ -0,0 +1,442 @@
+)
+
+const (
+	apkEnvKey     = "BITRISE_APK_PATH"
+	apkListEnvKey = "BITRISE_APK_PATH_LIST"
+
+	aabEnvKey     = "BITRISE_AAB_PATH"
+	aabListEnvKey = "BITRISE_AAB_PATH_LIST"
+
+	mappingFileEnvKey  = "BITRISE_MAPPING_PATH"
+	mappingFilePattern = "*build/*/mapping.txt"
+)
+
+type Result struct {
+	appArtifacts []gradle.Artifact
+	mappings     []gradle.Artifact
+}
+
+type InputParser interface {
+	Parse() (Input, error)
+}
+
+type envInputParser struct{}
+
+func NewInputParser() InputParser {
+	return envInputParser{}
+}
+
+func (envInputParser) Parse() (Input, error) {
+	var i Input
+	if err := stepconf.Parse(&i); err != nil {
+		return Input{}, err
+	}
+	return i, nil
+}
+
+type AndroidBuild struct {
+	stepInputParser InputParser
+}
+
+// TODO: bloat
+func NewAndroidBuild(stepInputParser InputParser) *AndroidBuild {
+	return &AndroidBuild{stepInputParser: stepInputParser}
+}
+
+func (a AndroidBuild) ProcessConfig() (Config, error) {
+	input, err := a.stepInputParser.Parse()
+	if err != nil {
+		return Config{}, nil
+	}
+	stepconf.Print(input)
+	return Config{
+		ProjectLocation: input.ProjectLocation,
+		AppPathPattern:  input.AppPathPattern,
+		Variant:         input.Variant,
+		Module:          input.Module,
+		AppType:         AppType(input.BuildType),
+		Arguments:       input.Arguments,
+		CacheLevel:      cache.Level(input.CacheLevel),
+		DeployDir:       input.DeployDir,
+	}, nil
+}
+
+func (a AndroidBuild) Run(cfg Config) (Result, error) {
+	gradleProject, err := gradle.NewProject(cfg.ProjectLocation)
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to open Gradle project: %s", err)
+	}
+
+	var buildTask *gradle.Task
+	if cfg.AppType == AppTypeAPK {
+		buildTask = gradleProject.GetTask("assemble")
+	} else {
+		buildTask = gradleProject.GetTask("bundle")
+	}
+
+	variants, err := buildTask.GetVariants()
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to fetch variants: %s", err)
+	}
+
+	filteredVariants, err := filterVariants(cfg.Module, cfg.Variant, variants)
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to find buildable variants: %s", err)
+	}
+
+	printVariants(variants, filteredVariants)
+
+	started := time.Now()
+
+	if err := executeGradleBuild(cfg, buildTask, filteredVariants); err != nil {
+		return Result{}, err
+	}
+
+	fmt.Println()
+	log.Infof("Export Artifacts:")
+
+	var appPathPatterns = strings.Split(cfg.AppPathPattern, "\n")
+	appArtifacts, err := getArtifacts(gradleProject, started, appPathPatterns, false)
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to find app artifacts: %v", err)
+	}
+
+	mappings, err := getArtifacts(gradleProject, started, []string{mappingFilePattern}, true)
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to find mapping files, error: %v", err)
+	}
+
+	printAppSearchInfo(appArtifacts, appPathPatterns)
+
+	return Result{
+		appArtifacts: appArtifacts,
+		mappings:     mappings,
+	}, nil
+}
+
+func (a AndroidBuild) Export(result Result, cfg Config) error {
```

Comment:
I think we should not pass the whole Config to the export.

- Result could have separate properties for apks and bundles, so that here we do not need the AppType.
- if len(filteredArtifacts) == 0 { could be done in Run, so we do not need AppPathPattern here.
- it seems DeployDir is the only remaining config we need here